File: core/MaximalSquare/MaximalSquare.java
package com.dbc;

import java.util.Arrays;

public class MaximalSquare {
    public int maximalSquare(char[][] matrix) {
        int[] dp = new int[matrix[0].length + 1];
        Arrays.fill(dp, 0);
        int res = 0, remain = 0;
        for (int row = 1; row <= matrix.length; row++) {
            for (int col = 1; col <= matrix[0].length; col++) {
                int temp = dp[col];
                if (matrix[row - 1][col - 1] == '0') dp[col] = 0;
                else dp[col] = Math.min(Math.min(dp[col - 1], temp), remain) + 1;
                remain = temp;
                res = Math.max(res, dp[col]);
            }
        }
        return res * res;
    }
}

File: core/MaximalSquare/MaximalSquare.py
from typing import List


def maximalSquare(matrix: List[List[str]]) -> int:
    rows, cols = len(matrix) + 1, len(matrix[0]) + 1
    res, dp, remain = 0, [0] * cols, 0
    for row in range(1, rows):
        for col in range(1, cols):
            temp = dp[col]
            if matrix[row - 1][col - 1] == "0":
                dp[col] = 0
            else:
                dp[col] = min(min(dp[col - 1], temp), remain) + 1
            remain = temp
            res = max(res, dp[col])

    return res ** 2





